sys/xtimer/xtimer_core.c

```diff
@@ -79,79 +66,64 @@ void xtimer_init(void)
-{
-    uint32_t before, after, long_value;
-
-    /* loop to cope with possible overflow of _xtimer_now() */
-    do {
-        before = _xtimer_now();
-        long_value = _long_cnt;
-        after = _xtimer_now();
-
-    } while (before > after);
-
-    *short_term = after;
-    *long_term = long_value;
+    _lltimer_ongoing = false;
 }
 
 uint64_t _xtimer_now64(void)
 {
-    uint32_t short_term, long_term;
-
-    _xtimer_now_internal(&short_term, &long_term);
+    /* time sensitive */
+    uint8_t state = irq_disable();
+    _xtimer_now();
+    irq_restore(state);
 
-    return ((uint64_t)long_term << 32) + short_term;
+    return _xtimer_current_time;
 }
 
 void _xtimer_set64(xtimer_t *timer, uint32_t offset, uint32_t long_offset)
 {
     DEBUG(" _xtimer_set64() offset=%" PRIu32 " long_offset=%" PRIu32 "\n", offset, long_offset);
-    if (!long_offset) {
-        /* timer fits into the short timer */
-        _xtimer_set(timer, (uint32_t)offset);
-    }
-    else {
-        int state = irq_disable();
-        if (_is_set(timer)) {
-            _remove(timer);
-        }
 
-        _xtimer_now_internal(&timer->target, &timer->long_target);
-        timer->target += offset;
-        timer->long_target += long_offset;
-        if (timer->target < offset) {
-            timer->long_target++;
-        }
-
-        _add_timer_to_long_list(&long_list_head, timer);
-        irq_restore(state);
-        DEBUG("xtimer_set64(): added longterm timer (long_target=%" PRIu32 " target=%" PRIu32 ")\n",
-              timer->long_target, timer->target);
-    }
-}
-
-void _xtimer_set(xtimer_t *timer, uint32_t offset)
-{
-    DEBUG("timer_set(): offset=%" PRIu32 " now=%" PRIu32 " (%" PRIu32 ")\n",
-          offset, xtimer_now().ticks32, _xtimer_lltimer_now());
     if (!timer->callback) {
         DEBUG("timer_set(): timer has no callback.\n");
         return;
     }
 
     xtimer_remove(timer);
 
-    if (offset < XTIMER_BACKOFF) {
+    if (!long_offset && offset < XTIMER_BACKOFF) {
+        /* timer fits into the short timer */
         _xtimer_spin(offset);
         _shoot(timer);
+        return;
+    }
+
+    /* time sensitive */
+    uint8_t state = irq_disable();
+    timer->offset = offset;
+    timer->long_offset = long_offset;
+    timer->start_time = _xtimer_now();
+
+    if (!long_offset) {
+        _add_timer_to_list(&timer_list_head, timer);
+
+        if (timer_list_head == timer) {
+            DEBUG("timer_set_absolute(): timer is new list head. updating lltimer.\n");
+            if (timer->offset <= _xtimer_lltimer_mask(0xFFFFFFFF)) {
+                /* schedule callback on next timer target time */
+                _lltimer_set(timer->start_time + offset);
+            }
+            else if (!_lltimer_ongoing) {
+                /* schedule callback after max_low_level_time/2 to detect a cycle */
+                _lltimer_set(timer->start_time + (_xtimer_lltimer_mask(0xFFFFFFFF)>>1));
+            }
```

Comment:
Part 1:
This is basically like the one under this. Only with ` timer->start_time` instead of `_xtimer_now()`. Why is it diffrent? Could it be one function that does this part and is called in both?

Reply:
Done.